### PR TITLE
refactor: add generic lazy-pages, native and substrate-ri interfaces

### PR DESCRIPTION
**Pull request by grishasobol**

### CODE REVIEW

File: core-processor/src/ext.rs [hunk 1]
```diff
@@ -16,6 +16,8 @@
 // You should have received a copy of the GNU General Public License
 // along with this program. If not, see <https://www.gnu.org/licenses/>.
 
+use core::marker::PhantomData;
+
```
Comment (breathx) [inline core-processor/src/ext.rs:20]:
```suggestion
```

File: core-processor/src/ext.rs [hunk 2]
```diff
@@ -300,21 +304,23 @@ impl BackendAllocSyscallError for AllocExtError {
     }
 }
 
-struct LazyGrowHandler {
+struct LazyGrowHandler<L: LazyPagesInterface> {
     old_mem_addr: Option<u64>,
     old_mem_size: WasmPagesAmount,
+    phantom: PhantomData<L>,
```
Comment (breathx) [inline core-processor/src/ext.rs:310]:
```suggestion
    _phantom: PhantomData<L>,
```

File: core-processor/src/ext.rs [hunk 3]
```diff
@@ -344,10 +350,11 @@ pub struct Ext {
     //
     // It's temporary field, used to solve `core-audit/issue#22`.
     outgoing_gasless: u64,
+    phantom: PhantomData<L>,
```
Comment (breathx) [inline core-processor/src/ext.rs:353]: _phantom

File: core-processor/src/ext.rs (hunk 1 above)
Comment (grishasobol) [reply core-processor/src/ext.rs]: done

File: core-processor/src/ext.rs [hunk 4]
```diff
@@ -300,21 +304,23 @@ impl BackendAllocSyscallError for AllocExtError {
...
 
-impl<Context> GrowHandler<Context> for LazyGrowHandler {
+impl<Context, L: LazyPagesInterface> GrowHandler<Context> for LazyGrowHandler<L> {
     fn before_grow_action(ctx: &mut Context, mem: &mut impl Memory<Context>) -> Self {
         // New pages allocation may change wasm memory buffer location.
         // So we remove protections from lazy-pages
         // and then in `after_grow_action` we set protection back for new wasm memory buffer.
         let old_mem_addr = mem.get_buffer_host_addr(ctx);
-        gear_lazy_pages_interface::remove_lazy_pages_prot(ctx, mem);
+        L::remove_lazy_pages_prot(ctx, mem);
```
[the remark this answers is not in the document]
Comment (grishasobol) [reply core-processor/src/ext.rs]: done

File: core-processor/src/ext.rs (hunk 3 above)
Comment (grishasobol) [reply core-processor/src/ext.rs]: done

**Pull request merged**